test_application/src/main.rs

```diff
@@ -0,0 +1,272 @@
+    let mut private_key_share: Option<KeySharePrivate> = None;
+    let mut aux_info_public: Option<Vec<AuxInfoPublic>> = None;
+    let mut aux_info_private: Option<AuxInfoPrivate> = None;
+    let mut presign_record: Option<PresignRecord> = None;
+
+    for (sub_protocol, sid) in next_action {
+        match sub_protocol {
+            SubProtocol::StartKeyGen => {
+                // TODO: Whats the point of making a stand-alone `KeygenParticipant`
+                // let key_gen = KeygenParticipant::new(config.id.clone(),
+                // config.other_ids.clone());
+
+                let participant: Participant<KeygenParticipant> =
+                    Participant::from_config(&config, sid, ());
+                let (public_key_shares_local, private_key_share_local) =
+                    worker_handle_subprotocol(participant, &other_workers, &outgoing)?;
+                outgoing.send(MessageFromWorker::SubProtocolEnded)?;
+                public_key_shares = Some(public_key_shares_local);
+                private_key_share = Some(private_key_share_local);
+            }
+            SubProtocol::StartAuxInfo => {
+                let participant: Participant<AuxInfoParticipant> =
+                    Participant::from_config(&config, sid, ());
+                // TODO: Aux info takes no arguments?!
+                let (aux_info_public_local, aux_info_private_local) =
+                    worker_handle_subprotocol(participant, &other_workers, &outgoing)?;
+                outgoing.send(MessageFromWorker::SubProtocolEnded)?;
+                aux_info_public = Some(aux_info_public_local);
+                aux_info_private = Some(aux_info_private_local);
+            }
+            SubProtocol::StartPresign => {
+                let input: PresignInput = PresignInput::new(
+                    aux_info_public.clone().unwrap(),
+                    aux_info_private.clone().unwrap(),
+                    public_key_shares.clone().unwrap(),
+                    private_key_share.clone().unwrap(),
+                )?;
+                let participant: Participant<PresignParticipant> =
+                    Participant::from_config(&config, sid, input);
+                let presign_record_local =
+                    worker_handle_subprotocol(participant, &other_workers, &outgoing)?;
+                outgoing.send(MessageFromWorker::SubProtocolEnded)?;
+                presign_record = Some(presign_record_local);
+            }
+            SubProtocol::Sign => {
+                // TODO Weird that we create a vacuous participant for signing.
+                let signature =
+                    Participant::<KeygenParticipant>::sign(presign_record.take().unwrap(), hash())?;
+                info!("Computed signature: {:?}", signature);
+                outgoing.send(MessageFromWorker::SubProtocolEnded)?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn hash() -> Sha256 {
```

Comment:
Is this how I'm supposed to generate a `Sha256` to pass to our sign function? I wasn't sure.

Reply:
I just looked more carefully at the `PresignRecord::sign` function and it actually calls `finalize` internally. I guess this was a design choice to force the caller to use a sha256 hash specifically, and not just a random 256 byte string.

Reply:
Conversation moved to #251